ssl/ssl_lib.c

```diff
@@ -6057,3 +6063,141 @@ int SSL_CTX_set0_tmp_dh_pkey(SSL_CTX *ctx, EVP_PKEY *dhpkey)
     ctx->cert->dh_tmp = dhpkey;
     return 1;
 }
+
+#ifndef OPENSSL_NO_COMP
+
+static int ssl_set_cert_comp_pref(uint16_t *array, size_t array_size, uint16_t *algs, int num_algs)
+{
+    int i, j = 0;
+
+    /* This will 0-terminate the array */
+    memset(array, 0, array_size);
+    /* include only those algorithms we support, ifnore the rest if the list fills */
+    for (i = 0; i < num_algs && j < TLSEXT_comp_cert_limit; i++) {
+#if defined(BROTLI)
+        if (algs[i] == TLSEXT_comp_cert_brotli)
+            array[j++] = algs[i];
+#endif
+#if defined(ZLIB)
+        if (algs[i] == TLSEXT_comp_cert_zlib)
+            array[j++] = algs[i];
+#endif
+#if defined(ZSTD)
+        if (algs[i] == TLSEXT_comp_cert_zstd)
+            array[j++] = algs[i];
+#endif
```

Comment:
Should we error on an unrecognised alg, or tolerate it?

Reply:
I would ignore it. A future version of OpenSSL might support additional algorithms, and if they were to use an older version, it should not error out; but accept what it knew about.